### PR TITLE
jsk_percetion: add ssd object detector

### PR DESCRIPTION
**Pull request by furushchev**
frame0000

For more detail: https://github.com/furushchev/jsk_recognition/blob/77d12bacb4ccb2c6451e4f025b7092042a416c05/doc/jsk_perception/nodes/ssd_object_detector.md

- [x] predict
- [x] train
- [x] documentation
- [x] sample
- [x] test code

### CODE REVIEW

File: jsk_perception/node_scripts/ssd_object_detector.py [hunk 1]
```diff
@@ -0,0 +1,176 @@
...
+
+    def image_cb(self, msg):
+        if self.profiling:
+            rospy.loginfo("callback start: incomming msg is %s msec behind" % ((rospy.Time.now() - msg.header.stamp).to_sec() * 1000.0))
+        tprev = time.time()
+        try:
+            # transform image to RGB, float, CHW
+            img = self.cv_bridge.imgmsg_to_cv2(msg, desired_encoding="rgb8")
+            img = np.asarray(img, dtype=np.float32)
+            img = img.transpose(2, 0, 1)  # (H, W, C) -> (C, H, W)
```
Comment (yuyu2172) [inline jsk_perception/node_scripts/ssd_object_detector.py]: I think you need to convert images to RGB. (CV2 returns BGR images)

Comment (yuyu2172) [reply jsk_perception/node_scripts/ssd_object_detector.py]: Sorry. This is my mistake.
It is RGB.

File: doc/jsk_perception/nodes/ssd_object_detector.md [hunk 2]
```diff
@@ -0,0 +1,181 @@
...
+    Weight files are generated on every 200 iterations by default. (e.g. `result/model_iter_400` as a weight file after 400 iterations)
+    You can monitor loss and stop training at anytime and resume with option `--resume`.
+
+    A generated weight file can be used as model file for `ssd_object_detector.py` by setting `~model_path` parameter as path to the weight file.
+
+## Reference
+
+```
+Wei Liu, et al. "SSD: Single shot multibox detector" ECCV 2016
+```
```
Comment (yuyu2172) [inline doc/jsk_perception/nodes/ssd_object_detector.md]: How about adding two references to the documentation.

[ChainerCV SSD code](https://github.com/chainer/chainercv/tree/master/examples/ssd)
[Image Labelling tools](https://github.com/yuyu2172/image-labelling-tool/tree/master/examples/ssd)

Comment (furushchev) [reply doc/jsk_perception/nodes/ssd_object_detector.md]: @yuyu2172 Sounds great! Updated docs :+1:

**Comment by furushchev** [comment]
Any other comments?
@k-okada @yuyu2172

**Comment by yuyu2172** [comment]
Looks great.
@k-okada  @wkentaro

File: jsk_perception/node_scripts/ssd_object_detector.py [hunk 3]
```diff
@@ -0,0 +1,176 @@
...
+
+        if self.profiling:
+            tcur = time.time()
+            rospy.loginfo("%s: elapsed %f msec" % ("callback end", (tcur-tprev)*1000))
+            tprev = tcur
+
+    def publish_bbox_image(self, img, bbox, label, score):
+        vis_bbox(img, bbox, label, score,
+                 label_names=self.label_names)
+        fig = plt.gcf()
```
Comment (knorth55) [inline jsk_perception/node_scripts/ssd_object_detector.py]: i think we do not need to plot in this node.

Comment (furushchev) [reply jsk_perception/node_scripts/ssd_object_detector.py]: @knorth55 Why?

Comment (knorth55) [reply jsk_perception/node_scripts/ssd_object_detector.py]: faster_rcnn nodes also draw boxes and labels, so I think we can make a common node like `draw_boxes`

Comment (furushchev) [reply jsk_perception/node_scripts/ssd_object_detector.py]: @knorth55 Thanks! I created pull request to generalize visualization part. https://github.com/jsk-ros-pkg/jsk_recognition/pull/2205

Comment (furushchev) [reply jsk_perception/node_scripts/ssd_object_detector.py]: - #2205 was merged.
- Updated to use `draw_rects` for visualization
- Remove visualization function from `ssd_object_detector.py` node
- Updated docs

File: jsk_perception/node_scripts/ssd_train_dataset.py [hunk 4]
```diff
@@ -0,0 +1,244 @@
+#!/usr/bin/env python
```
Comment (knorth55) [inline jsk_perception/node_scripts/ssd_train_dataset.py]: I think this file is not node.
It should be placed in `scripts/` or somewhere

Comment (wkentaro) [reply jsk_perception/node_scripts/ssd_train_dataset.py]: +1

Comment (furushchev) [reply jsk_perception/node_scripts/ssd_train_dataset.py]: @knorth55 @wkentaro True! Moved to `scripts`.

**Comment by knorth55** [comment]
@furushchev  test failing.
```
Full test results for 'jsk_perception/test_results/jsk_perception/rosunit-test_ssd_object_detector.xml'

-------------------------------------------------
<?xml version="1.0" encoding="utf-8"?>

<testsuite errors="1" failures="0" name="unittest.suite.TestSuite" tests="1" time="59.635">

  <testcase classname="__main__.TestTopicPublished" name="test_published" time="59.6345">

    <error type="ValueError">topic parameter 'data_class' is not initialized

  File "/usr/lib/python2.7/unittest/case.py", line 331, in run

    testMethod()

  File "/opt/ros/indigo/share/jsk_tools/src/test_topic_published.py", line 76, in test_published

    checker = PublishChecker(topic_name, timeout)

  File "/opt/ros/indigo/share/jsk_tools/src/test_topic_published.py", line 27, in __init__

    self.sub = rospy.Subscriber(topic_name, msg_class, self._callback)

  File "/opt/ros/indigo/lib/python2.7/dist-packages/rospy/topics.py", line 533, in __init__

    super(Subscriber, self).__init__(name, data_class, Registration.SUB)

  File "/opt/ros/indigo/lib/python2.7/dist-packages/rospy/topics.py", line 136, in __init__

    raise ValueError("topic parameter 'data_class' is not initialized")

    </error>

  </testcase>

  <system-out><![CDATA[]]></system-out>

  <system-err><![CDATA[WARNING: topic [/ssd_object_detector/output/rect] does not appear to be published yet

]]></system-err>

</testsuite>
```

```
Full test results for 'jsk_perception/test_results/jsk_perception/rostest-test_ssd_object_detector.xml'

-------------------------------------------------

<?xml version="1.0" encoding="utf-8"?>

<testsuite errors="1" failures="0" name="unittest.suite.TestSuite" tests="1" time="245.203">

  <testcase classname="rostest.runner.RosTest" name="testtest_ssd_object_detector" time="245.2023">

    <error type="RLTestTimeoutException">max time [60.0s] allotted for test [test_ssd_object_detector] of type [jsk_tools/test_topic_published.py]

  File "/usr/lib/python2.7/unittest/case.py", line 331, in run

    testMethod()

  File "/opt/ros/indigo/lib/python2.7/dist-packages/rostest/runner.py", line 148, in fn

    self.test_parent.run_test(test)

  File "/opt/ros/indigo/lib/python2.7/dist-packages/rostest/rostest_parent.py", line 132, in run_test

    return self.runner.run_test(test)

  File "/opt/ros/indigo/lib/python2.7/dist-packages/roslaunch/launch.py", line 681, in run_test

    (test.time_limit, test.test_name, test.package, test.type))

    </error>

  </testcase>

  <system-out><![CDATA[[ROSTEST]setup[/workspace/ros/ws_jsk_recognition/src/jsk_recognition/jsk_perception/test/ssd_object_detector.test] run_id[1b48436c-b891-11e7-a7d9-0242ac110005] starting

?[1mstarted roslaunch server http://d50497e39805:45053/?[0m



SUMMARY

========



PARAMETERS

 * /image_publisher/file_name: /workspace/ros/ws...

 * /image_publisher/publish_info: False

 * /rosdistro: indigo

 * /rosversion: 1.11.21

 * /ssd_object_detector/gpu: -1

 * /test_ssd_object_detector/timeout_0: 120

 * /test_ssd_object_detector/timeout_1: 120

 * /test_ssd_object_detector/timeout_2: 120

 * /test_ssd_object_detector/topic_0: /ssd_object_detec...

 * /test_ssd_object_detector/topic_1: /ssd_object_detec...

 * /test_ssd_object_detector/topic_2: /ssd_object_detec...



NODES

  /

    image_publisher (jsk_perception/image_publisher.py)

    ssd_object_detector (jsk_perception/ssd_object_detector.py)



[ROSTEST]setup[/workspace/ros/ws_jsk_recognition/src/jsk_recognition/jsk_perception/test/ssd_object_detector.test] run_id[1b48436c-b891-11e7-a7d9-0242ac110005] done

[ROSTEST]Running test [test_ssd_object_detector]

?[1mROS_MASTER_URI=http://d50497e39805:34590/?[0m

?[1msetting /run_id to 1b48436c-b891-11e7-a7d9-0242ac110005?[0m

?[1mprocess[rosout-1]: started with pid [52296]?[0m

started core service [/rosout]

?[1mprocess[image_publisher-2]: started with pid [52313]?[0m

?[1mprocess[ssd_object_detector-3]: started with pid [52314]?[0m

[ROSTEST]running test test_ssd_object_detector

?[1mprocess[test_ssd_object_detector-4]: started with pid [52315]?[0m

[ROSTEST]test [test_ssd_object_detector] failed, retrying. Retries left: 2

[ROSTEST]tearDown[/workspace/ros/ws_jsk_recognition/src/jsk_recognition/jsk_perception/test/ssd_object_detector.test]

shutting down processing monitor...

[test_ssd_object_detector-4] killing on exit

[image_publisher-2] killing on exit

[rosout-1] killing on exit

... shutting down processing monitor complete

[ROSTEST]rostest teardown /workspace/ros/ws_jsk_recognition/src/jsk_recognition/jsk_perception/test/ssd_object_detector.test complete

[ROSTEST]setup[/workspace/ros/ws_jsk_recognition/src/jsk_recognition/jsk_perception/test/ssd_object_detector.test] run_id[3ffbc8b4-b891-11e7-a7d9-0242ac110005] starting

?[1mstarted roslaunch server http://d50497e39805:56871/?[0m



SUMMARY

========



PARAMETERS

 * /image_publisher/file_name: /workspace/ros/ws...

 * /image_publisher/publish_info: False

 * /rosdistro: indigo

 * /rosversion: 1.11.21

 * /ssd_object_detector/gpu: -1

 * /test_ssd_object_detector/timeout_0: 120

 * /test_ssd_object_detector/timeout_1: 120

 * /test_ssd_object_detector/timeout_2: 120

 * /test_ssd_object_detector/topic_0: /ssd_object_detec...

 * /test_ssd_object_detector/topic_1: /ssd_object_detec...

 * /test_ssd_object_detector/topic_2: /ssd_object_detec...



NODES

  /

    image_publisher (jsk_perception/image_publisher.py)

    ssd_object_detector (jsk_perception/ssd_object_detector.py)



[ROSTEST]setup[/workspace/ros/ws_jsk_recognition/src/jsk_recognition/jsk_perception/test/ssd_object_detector.test] run_id[3ffbc8b4-b891-11e7-a7d9-0242ac110005] done

[ROSTEST]Running test [test_ssd_object_detector]

?[1mROS_MASTER_URI=http://d50497e39805:46344/?[0m

?[1msetting /run_id to 3ffbc8b4-b891-11e7-a7d9-0242ac110005?[0m

?[1mprocess[rosout-5]: started with pid [53045]?[0m

started core service [/rosout]

?[1mprocess[image_publisher-6]: started with pid [53062]?[0m

?[1mprocess[ssd_object_detector-7]: started with pid [53063]?[0m

[ROSTEST]removing previous test results file [/workspace/ros/ws_jsk_recognition/build/jsk_perception/test_results/jsk_perception/rosunit-test_ssd_object_detector.xml]

[ROSTEST]running test test_ssd_object_detector

?[1mprocess[test_ssd_object_detector-8]: started with pid [53075]?[0m

[ROSTEST]test [test_ssd_object_detector] failed, retrying. Retries left: 1

[ROSTEST]tearDown[/workspace/ros/ws_jsk_recognition/src/jsk_recognition/jsk_perception/test/ssd_object_detector.test]

shutting down processing monitor...

[test_ssd_object_detector-8] killing on exit

[image_publisher-6] killing on exit

[rosout-5] killing on exit

... shutting down processing monitor complete

[ROSTEST]rostest teardown /workspace/ros/ws_jsk_recognition/src/jsk_recognition/jsk_perception/test/ssd_object_detector.test complete

[ROSTEST]setup[/workspace/ros/ws_jsk_recognition/src/jsk_recognition/jsk_perception/test/ssd_object_detector.test] run_id[64906fc2-b891-11e7-a7d9-0242ac110005] starting

?[1mstarted roslaunch server http://d50497e39805:43674/?[0m



SUMMARY

========



PARAMETERS

 * /image_publisher/file_name: /workspace/ros/ws...

 * /image_publisher/publish_info: False

 * /rosdistro: indigo

 * /rosversion: 1.11.21

 * /ssd_object_detector/gpu: -1

 * /test_ssd_object_detector/timeout_0: 120

 * /test_ssd_object_detector/timeout_1: 120

 * /test_ssd_object_detector/timeout_2: 120

 * /test_ssd_object_detector/topic_0: /ssd_object_detec...

 * /test_ssd_object_detector/topic_1: /ssd_object_detec...

 * /test_ssd_object_detector/topic_2: /ssd_object_detec...



NODES

  /

    image_publisher (jsk_perception/image_publisher.py)

    ssd_object_detector (jsk_perception/ssd_object_detector.py)



[ROSTEST]setup[/workspace/ros/ws_jsk_recognition/src/jsk_recognition/jsk_perception/test/ssd_object_detector.test] run_id[64906fc2-b891-11e7-a7d9-0242ac110005] done

[ROSTEST]Running test [test_ssd_object_detector]

?[1mROS_MASTER_URI=http://d50497e39805:42711/?[0m

?[1msetting /run_id to 64906fc2-b891-11e7-a7d9-0242ac110005?[0m

?[1mprocess[rosout-9]: started with pid [53795]?[0m

started core service [/rosout]

?[1mprocess[image_publisher-10]: started with pid [53812]?[0m

?[1mprocess[ssd_object_detector-11]: started with pid [53813]?[0m

[ROSTEST]removing previous test results file [/workspace/ros/ws_jsk_recognition/build/jsk_perception/test_results/jsk_perception/rosunit-test_ssd_object_detector.xml]

[ROSTEST]running test test_ssd_object_detector

?[1mprocess[test_ssd_object_detector-12]: started with pid [53822]?[0m

[ROSTEST]test [test_ssd_object_detector] failed, retrying. Retries left: 0

[ROSTEST]tearDown[/workspace/ros/ws_jsk_recognition/src/jsk_recognition/jsk_perception/test/ssd_object_detector.test]

shutting down processing monitor...

[test_ssd_object_detector-12] killing on exit

[image_publisher-10] killing on exit

[rosout-9] killing on exit

... shutting down processing monitor complete

[ROSTEST]rostest teardown /workspace/ros/ws_jsk_recognition/src/jsk_recognition/jsk_perception/test/ssd_object_detector.test complete

[ROSTEST]setup[/workspace/ros/ws_jsk_recognition/src/jsk_recognition/jsk_perception/test/ssd_object_detector.test] run_id[88e80970-b891-11e7-a7d9-0242ac110005] starting

?[1mstarted roslaunch server http://d50497e39805:39969/?[0m



SUMMARY

========



PARAMETERS

 * /image_publisher/file_name: /workspace/ros/ws...

 * /image_publisher/publish_info: False

 * /rosdistro: indigo

 * /rosversion: 1.11.21

 * /ssd_object_detector/gpu: -1

 * /test_ssd_object_detector/timeout_0: 120

 * /test_ssd_object_detector/timeout_1: 120

 * /test_ssd_object_detector/timeout_2: 120

 * /test_ssd_object_detector/topic_0: /ssd_object_detec...

 * /test_ssd_object_detector/topic_1: /ssd_object_detec...

 * /test_ssd_object_detector/topic_2: /ssd_object_detec...



NODES

  /

    image_publisher (jsk_perception/image_publisher.py)

    ssd_object_detector (jsk_perception/ssd_object_detector.py)



[ROSTEST]setup[/workspace/ros/ws_jsk_recognition/src/jsk_recognition/jsk_perception/test/ssd_object_detector.test] run_id[88e80970-b891-11e7-a7d9-0242ac110005] done

[ROSTEST]Running test [test_ssd_object_detector]

?[1mROS_MASTER_URI=http://d50497e39805:33742/?[0m

?[1msetting /run_id to 88e80970-b891-11e7-a7d9-0242ac110005?[0m

?[1mprocess[rosout-13]: started with pid [54546]?[0m

started core service [/rosout]

?[1mprocess[image_publisher-14]: started with pid [54563]?[0m

?[1mprocess[ssd_object_detector-15]: started with pid [54564]?[0m

[ROSTEST]removing previous test results file [/workspace/ros/ws_jsk_recognition/build/jsk_perception/test_results/jsk_perception/rosunit-test_ssd_object_detector.xml]

[ROSTEST]running test test_ssd_object_detector

?[1mprocess[test_ssd_object_detector-16]: started with pid [54575]?[0m

[ROSTEST]tearDown[/workspace/ros/ws_jsk_recognition/src/jsk_recognition/jsk_perception/test/ssd_object_detector.test]

shutting down processing monitor...

[test_ssd_object_detector-16] killing on exit

[image_publisher-14] killing on exit

[rosout-13] killing on exit

... shutting down processing monitor complete

[ROSTEST]rostest teardown /workspace/ros/ws_jsk_recognition/src/jsk_recognition/jsk_perception/test/ssd_object_detector.test complete

]]></system-out>

  <system-err><![CDATA[?[31m[ssd_object_detector-3] process has died [pid 52314, exit code 1, cmd /workspace/ros/ws_jsk_recognition/src/jsk_recognition/jsk_perception/node_scripts/ssd_object_detector.py ~input:=image_publisher/output __name:=ssd_object_detector __log:=/workspace/jsk-ros-pkg/jsk_recognition/log/1b48436c-b891-11e7-a7d9-0242ac110005/ssd_object_detector-3.log].

log file: /workspace/jsk-ros-pkg/jsk_recognition/log/1b48436c-b891-11e7-a7d9-0242ac110005/ssd_object_detector-3*.log?[0m

[ROSTEST]test [test_ssd_object_detector] timed out

?[31m[ssd_object_detector-7] process has died [pid 53063, exit code 1, cmd /workspace/ros/ws_jsk_recognition/src/jsk_recognition/jsk_perception/node_scripts/ssd_object_detector.py ~input:=image_publisher/output __name:=ssd_object_detector __log:=/workspace/jsk-ros-pkg/jsk_recognition/log/3ffbc8b4-b891-11e7-a7d9-0242ac110005/ssd_object_detector-7.log].

log file: /workspace/jsk-ros-pkg/jsk_recognition/log/3ffbc8b4-b891-11e7-a7d9-0242ac110005/ssd_object_detector-7*.log?[0m

[ROSTEST]test [test_ssd_object_detector] timed out

?[31m[ssd_object_detector-11] process has died [pid 53813, exit code 1, cmd /workspace/ros/ws_jsk_recognition/src/jsk_recognition/jsk_perception/node_scripts/ssd_object_detector.py ~input:=image_publisher/output __name:=ssd_object_detector __log:=/workspace/jsk-ros-pkg/jsk_recognition/log/64906fc2-b891-11e7-a7d9-0242ac110005/ssd_object_detector-11.log].

log file: /workspace/jsk-ros-pkg/jsk_recognition/log/64906fc2-b891-11e7-a7d9-0242ac110005/ssd_object_detector-11*.log?[0m

[ROSTEST]test [test_ssd_object_detector] timed out

?[31m[ssd_object_detector-15] process has died [pid 54564, exit code 1, cmd /workspace/ros/ws_jsk_recognition/src/jsk_recognition/jsk_perception/node_scripts/ssd_object_detector.py ~input:=image_publisher/output __name:=ssd_object_detector __log:=/workspace/jsk-ros-pkg/jsk_recognition/log/88e80970-b891-11e7-a7d9-0242ac110005/ssd_object_detector-15.log].

log file: /workspace/jsk-ros-pkg/jsk_recognition/log/88e80970-b891-11e7-a7d9-0242ac110005/ssd_object_detector-15*.log?[0m

]]></system-err>

</testsuite>
```

**Comment by furushchev** [comment]
@knorth55 Thx for report. Extended time limit.

**Comment by k-okada** [comment]
```
Full test results for 'jsk_perception/test_results/jsk_perception/rosunit-test_ssd_object_detector.xml'
-------------------------------------------------
<?xml version="1.0" encoding="utf-8"?>
<testsuite errors="1" failures="0" name="unittest.suite.TestSuite" tests="1" time="59.737">
  <testcase classname="__main__.TestTopicPublished" name="test_published" time="59.7372">
    <error type="ValueError">topic parameter 'data_class' is not initialized
  File "/usr/lib/python2.7/unittest/case.py", line 331, in run
    testMethod()
  File "/opt/ros/indigo/share/jsk_tools/src/test_topic_published.py", line 76, in test_published
    checker = PublishChecker(topic_name, timeout)
  File "/opt/ros/indigo/share/jsk_tools/src/test_topic_published.py", line 27, in __init__
    self.sub = rospy.Subscriber(topic_name, msg_class, self._callback)
  File "/opt/ros/indigo/lib/python2.7/dist-packages/rospy/topics.py", line 533, in __init__
    super(Subscriber, self).__init__(name, data_class, Registration.SUB)
  File "/opt/ros/indigo/lib/python2.7/dist-packages/rospy/topics.py", line 136, in __init__
    raise ValueError("topic parameter 'data_class' is not initialized")
    </error>
  </testcase>
  <system-out><![CDATA[]]></system-out>
  <system-err><![CDATA[WARNING: topic [/ssd_object_detector/output/rect] does not appear to be published yet
]]></system-err>
</testsuite>
-------------------------------------------------
Skipping "jsk_perception/test_results/jsk_perception/rosunit-test_people_pose_estimation_2d.xml": no element found: line 2, column 0
Full test results for 'jsk_perception/test_results/jsk_perception/rostest-test_ssd_object_detector.xml'
-------------------------------------------------
<?xml version="1.0" encoding="utf-8"?>
<testsuite errors="1" failures="0" name="unittest.suite.TestSuite" tests="1" time="245.011">
  <testcase classname="rostest.runner.RosTest" name="testtest_ssd_object_detector" time="245.0110">
    <error type="RLTestTimeoutException">max time [60.0s] allotted for test [test_ssd_object_detector] of type [jsk_tools/test_topic_published.py]
  File "/usr/lib/python2.7/unittest/case.py", line 331, in run
    testMethod()
  File "/opt/ros/indigo/lib/python2.7/dist-packages/rostest/runner.py", line 148, in fn
    self.test_parent.run_test(test)
  File "/opt/ros/indigo/lib/python2.7/dist-packages/rostest/rostest_parent.py", line 132, in run_test
    return self.runner.run_test(test)
  File "/opt/ros/indigo/lib/python2.7/dist-packages/roslaunch/launch.py", line 681, in run_test
    (test.time_limit, test.test_name, test.package, test.type))
    </error>
  </testcase>
  <system-out><![CDATA[[ROSTEST]setup[/workspace/ros/ws_jsk_recognition/src/jsk_recognition/jsk_perception/test/ssd_object_detector.test] run_id[34629c5c-cf35-11e7-83eb-0242ac110005] starting
?[1mstarted roslaunch server http://24b761d2ec96:56993/?[0m
```

**Comment by furushchev** [comment]
@k-okada Thanks for report. Updated.

**Comment by furushchev** [comment]
Test passed.

**Comment by k-okada** [comment]
@furushchev thank you!

File: jsk_perception/node_scripts/ssd_train_dataset.py (hunk 4 above)
Comment (furushchev) [reply jsk_perception/node_scripts/ssd_train_dataset.py]: Oh no. It is still in `node_scripts` directory. Maybe I forgot to push this commit? If you mind, I'll create another pull request later, sorry.

Comment (knorth55) [reply jsk_perception/node_scripts/ssd_train_dataset.py]: @furushchev I moved it to `script` in #2305

**Pull request merged**